### DISCUSSION TITLE
libpag的线程池是否可以优化

### DISCUSSION THREAD

**Discussion by peter100u** [Tencent/libpag#2009] [Q&A] [answered]
在大量请求pag的时候，有些手机会ANR, 看到profiler中的数据，这里8个线程，这里是线程池吗，是否可以进行线程优化，而且这些线程并销毁，一直存在，导致线程访问文件句柄过大，导致低端设备anr
![image](https://github.com/Tencent/libpag/assets/14010579/fa1612b4-7817-4352-87c5-b934a401a749)

**Comment by kevingpqi123** [comment, marked as answer]
感觉 ANR 和这里的线程数量关系不大，在 pag 中，内部维护有一个线程池，最大线程数量：static const int MaxThreads = CPUCores > 16 ? 16 : CPUCores; 线程数量和 pag 同时渲染的数量强相关，且使用完会及时进行销毁
关于 ANR，有以下三个分析方向：
1、不是所有的 pag 文件的渲染性能都是一样的，pag 文件渲染的性能和素材复杂度强相关，其次就是和机型的计算能力，最直接的优化是素材，在中低端机上尽量使用纯矢量的素材

2、卡顿还有一个原因可能是在 PAGView 播放的过程中直接销毁，由于 pag 是内部线程安全， 播放过程中每帧渲染的 flush 方法相对耗时，一种解决方案是页面销毁前提前调用 PAGView 的 stop 方法，尽可能早的终止播放，避免销毁时的方法调用等待播放时的 flush 方法

3、第三种优化方法：如果不涉及全屏播放、文本编辑或占位图替换，可以升级到 PAG 4.2 版本，切换使用 PAGImageView， PAGImageView 支持将渲染结果缓存至本地，缓存完成后后续使用的是缓存数据

**Comment by peter100u** [comment]
好的，感谢回复
